pages/form/index.tsx

```diff
@@ -0,0 +1,232 @@
+  return !passwordConfirm
+    ? "Confirm Password is required."
+    : state["password"].value !== passwordConfirm
+    ? "Passwords do not match."
+    : undefined;
+};
+
+/**
+ * Constants
+ */
+
+const registerFormFieldSpec: FieldSpec = {
+  email: {
+    type: "text",
+    label: "Email",
+    placeholder: "email@example.com",
+    validationFn: validateEmail,
+  },
+  firstName: {
+    label: "First Name",
+    type: "text",
+    placeholder: "Enter your First Name here",
+    validationFn: validateFirstName,
+  },
+  lastName: {
+    label: "Last Name",
+    type: "text",
+    placeholder: "Enter your Last Name here",
+    validationFn: validateLastName,
+  },
+  password: {
+    label: "Password",
+    type: "password",
+    placeholder: "Enter your Password here",
+    validationFn: validatePassword,
+  },
+  passwordConfirm: {
+    label: "Confirm Password",
+    type: "password",
+    placeholder: "Confirm your Password",
+    validationFn: validatePasswordConfirm,
+  },
+};
+
+const initialFieldState: FieldState = {
+  value: "",
+  error: undefined,
+};
+
+/**
+ * Types
+ */
+
+type FieldState = {
+  value: string;
+  error: string | undefined;
+};
+
+type FormState = {
+  [fieldName: string]: FieldState;
+};
+
+type FieldSpec = {
+  [fieldKey: string]: {
+    type: "text" | "password";
+    label: string;
+    placeholder?: string;
+    validationFn: (input: string, state: FormState) => string | undefined;
+  };
+};
+
+/**
+ * Custom Hook
+ */
+
+function useForm(fieldSpec: FieldSpec) {
+  const initialState: FormState = Object.fromEntries(
+    Object.keys(fieldSpec).map((fieldKey) => {
+      return [fieldKey, initialFieldState];
+    })
+  );
+
+  const [formState, setFormState] = useState(initialState);
+
+  const updateFieldValue = (fieldKey: string, newValue: string) => {
+    setFormState({
+      ...formState,
+      [fieldKey]: { ...formState[fieldKey], value: newValue },
+    });
+  };
+
+  const updateFormState = (
+    errors: [string, string | undefined][]
+  ): FormState => {
+    const callback = (
+      previousState: FormState,
+      [key, reason]: [key: string, reason: string | undefined]
+    ): FormState => {
+      return {
+        ...previousState,
+        [key]: { ...previousState[key], error: reason },
+      };
+    };
+    return errors.reduce(callback, formState);
+  };
+
+  return {
+    initialState,
+    formState,
+    setFormState,
+    updateFieldValue,
+    updateFormState,
+  };
+}
+
+/**
+ * Form Component
+ */
+
+const Form = ({ fieldSpec }: { fieldSpec: FieldSpec }) => {
+  const {
+    initialState,
+    formState,
+    setFormState,
+    updateFieldValue,
+    updateFormState,
+  } = useForm(fieldSpec);
+
+  const submitForm = async (validForm: FormState) => {
+    const result = await axios.post("/api/register", validForm);
+    console.log(result);
+    setFormState(initialState);
+  };
+
+  const handleSubmit = async (event: FormEvent) => {
```

Comment:
is this where it should belong? :wink:

Reply:
Hint: look at the variables inside `handleSubmit` 
are there any variables that only come from `Form`?

Reply:
I think it should belong to the place use `Form` component. Is it right?

Reply:
Let's take a look
- `event` - from event handler argument
- `fieldSpec` - from `useForm`
- `formState` - from `useForm`
- `updateFormState` from `useForm`
- `setFormState` from `useForm`
- `submitForm` from `Form`
- everything else are local variables

so.....what can you do?

Reply:
I got it! I've put all of them into useForm, but not sure if `submitForm` should be in `useForm` as well OR be passed as argument from `Form`.